Seminars/Sem.01/Pract.01/task03.cpp

```diff
@@ -0,0 +1,68 @@
+#include <iostream>
+
+using namespace std;
+
+int **createMatrix(const int rows, const int cols) {
+    int **matrix = new int *[rows];
+    for (int i = 0; i < rows; i++) {
+        matrix[i] = new int[cols];
+    }
+    return matrix;
+}
+
+void inputMatrix(int **matrix, const int rows, const int cols) {
+    for (int i = 0; i < rows; i++) {
+        for (int j = 0; j < cols; j++) {
+            cout << "Enter the element at position (" << i << ", " << j << "): ";
+            cin >> matrix[i][j];
+        }
+    }
+}
+
+int **transposeMatrix(int **matrix, const int rows, const int cols) {
```

Comment:
Should we use const?

Reply:
No, the function call `transposeMatrix(matrix, rows, cols)` doesn't recognize `int **transposeMatrix(const int **matrix, const int rows, const int cols)` because the passed parameter `matrix` isn't `const int**`.

Reply:
If it's just `const int** matrix` only the first pointer is constant and the function still can change the value of `**matrix` so in order to prevent that we need to add `const` before the second pointer?

Reply:
const int** matrix is a pointer -> pointer -> const int , so no pointer is constant and that is the actual issue.
const int* const* is a pointer -> const pointer -> const int, thus we make the value and the pointer "safe"(const <-> non-modifiable).

A little bit more:
Because converting Foo** → const Foo** would be invalid and dangerous.
C++ allows the (safe) conversion Foo* → Foo const*, but gives an error if you try to implicitly convert Foo** → const Foo**.
The reason the conversion from Foo** → const Foo** is dangerous is that it would let you silently and accidentally modify a const Foo object without a cast.

Reply:
Ok, I'll fix it.